### PR TITLE
Optimize kernel selector

### PR DESCRIPTION
**Pull request by ahnyoung-paul**
### Details:
 - *Optimize kernel selector*
 - *...*

### Tickets:
 - *ticket-id*

### CODE REVIEW

File: src/plugins/intel_gpu/src/graph/primitive_inst.cpp [hunk 1]
```diff
@@ -127,8 +127,18 @@ void primitive_inst::realloc_if_needed() {
 
 void primitive_inst::update_impl() {
     if (!_node.is_type<data>() && !(_node.is_type<mutable_data>() && _node.get_dependencies().empty())) {
-        _impl = std::move(_node.type()->choose_impl(_node));
-        _network.get_program()->compile();
+        auto layout_key = _node.get_output_layout().to_string();
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/primitive_inst.cpp:130]: I think input layout should be used as key.

File: src/plugins/intel_gpu/src/graph/program_node.cpp [hunk 2]
```diff
@@ -32,6 +32,8 @@ program_node::program_node(std::shared_ptr<primitive> prim, program& prog)
     : desc(prim), myprog(prog), org_id(prim->id) {
     if (prim)
         output_layout.data_padding = prim->output_padding;
+
+    primitive_impl_caches = std::make_shared<LRUCache<std::string, std::shared_ptr<primitive_impl>>>(4 * sizeof(std::shared_ptr<primitive_impl>));
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/program_node.cpp:36]: Lets make this "4" to be more understandable & controllable, i.e.g, declare it as a const variable at least.

File: src/plugins/intel_gpu/src/graph/primitive_inst.cpp [hunk 3]
```diff
@@ -127,8 +127,18 @@ void primitive_inst::realloc_if_needed() {
...
+        auto caches = _node.get_primitive_impl_caches();
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/primitive_inst.cpp:131]: There is one "cache" for a primitive. So "caches" seems to be misread.

File: src/plugins/intel_gpu/src/graph/primitive_inst.cpp (hunk 1 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/graph/primitive_inst.cpp:130]: I missed it, I will update layout key. thanks

File: src/plugins/intel_gpu/src/graph/primitive_inst.cpp [hunk 4]
```diff
@@ -127,8 +127,18 @@ void primitive_inst::realloc_if_needed() {
...
+        bool is_new_impl = false;
+        std::tie(_impl, is_new_impl) = caches->get(layout_key, [&]() {
+            cldnn::LRUCache<std::string, std::shared_ptr<cldnn::primitive_impl>>::CacheEntry new_entry;
+            new_entry.data = std::move(_node.type()->choose_impl(_node));
+            new_entry.size = sizeof(new_entry.data);
+            return new_entry;
+        });
+        if (is_new_impl) {
+            _network.get_program()->compile();
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/primitive_inst.cpp:140]: Umm There might be such a case that: 
1) there is a impl in a cache 
2) however that impl is evicted from kernels_cache::in_memory_cache due to the size limit 
3) Then we will be recreate the impl

hmm lets reconsider this flow

File: src/plugins/intel_gpu/src/graph/primitive_inst.cpp (hunk 1 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/graph/primitive_inst.cpp]: Fixed it

File: src/plugins/intel_gpu/src/graph/program_node.cpp (hunk 2 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/graph/program_node.cpp]: Fixed it.

File: src/plugins/intel_gpu/src/graph/primitive_inst.cpp (hunk 3 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/graph/primitive_inst.cpp]: Fixed it.

**Pull request closed without merging**